### ISSUE TITLE
Adding relay not adding properly

### ISSUE THREAD

**Issue by fishcharlie** [damus-io/damus#1849]
### Current Result

- After tapping `Add relay` panel doesn't dismiss (no indication that it added successfully)
- Relay doesn't show up in list without dismissing relay page and going back in
- Relay doesn't persist between launches

### Expected Result

When tapping `Add relay` I expect the panel to dismiss and the relay to show up in the list and persist between force quits.

### Steps to reproduce

From my perspective:

1. Open `My Relays`
2. Tap `Add relay`
3. Type in relay URL
4. Tap `Add relay`
5. Notice that nothing happens (panel doesn't dismiss, gives the indication that it didn't work)
6. Tap `Add relay` again
7. Notice that duplicate relay message is shown
8. Close relay
9. Scroll through relays list
10. Notice that relay doesn't show up
11. Go back to home page
12. Reopen relays page
13. Notice that relay now shows up
14. Force quit app
15. Relaunch app
16. Reopen relays page
17. Notice that newly added relay no longer shows up

Obviously I wouldn't be surprised if more steps took place prior to this to get damus into this state, but I can not identify what those would be.

### Video Recording

https://github.com/damus-io/damus/assets/860375/478c6a13-39a1-401c-a2ca-2e6bcd1f250e

### Version

**Damus**:

1.6 (28) 2585a375ab0a
TestFlight

Also observed on a2cac142 (development build I was using).

**iOS**:

17.2.1
iPhone 15 Pro Max

### Other Information

- I'm not sure if this is multiple issues. I kinda get the feeling this is a single root issue, which is why I created a single ticket. If it's determined to be multiple issues we can break it into multiple tickets.

**Comment by fishcharlie** [comment]
Notes:

I updated to the lasted commit before doing my testing.

- In `AddRelayView.swift` line 109 the `return` of that guard else is getting hit. Which means `add_relay` is returning nil
    - This explains why the view isn't dismissing
- Tracing that further back `Contacts.swift` line 208 is getting hit: `return nil` from the else of the guard `relays.index(forKey: relay) == nil`. Which means `relays.index(forKey: relay)` is not nil
- In that `add_relay` function the relay I'm trying to add exists in both `relays` & `current_relays` (`state.pool.our_descriptors`)
- Which makes sense because on line 95 of `AddRelayView.swift` it's calling `try state.pool.add_relay(descriptor)` which appends it to `self.relays` at the end of that function.
- The last person to update any of this code was @jb55 as part of 14a3a35, and before that it looks to have been @danieldaquino as part of 768ab3e

I personally don't see a path where this should ever succeed since we are adding the relay to the pool before we are calling Contacts.swift `add_relay` (which expects the relay to not have been added to the pool).

Problem is I don't know what the happy path is here. I'm not quite sure what the purpose is for having both Contacts.swift `add_relay` & `state.pool.add_relay`? I'm also not sure edge case ensuring `relays.index(forKey: relay) == nil` achieves?

I'd be happy to keep looking into this and fix it if someone gives me some guidance on that. Otherwise, hopefully these notes can help someone with more knowledge than me fix it.

**Comment by alltheseas** [comment]
I cannot recreate. 

No issues adding the relay on my end. Same iOS, Damus versions.

**Comment by danieldaquino** [comment]
Hi @fishcharlie! 

> I'm not quite sure what the purpose is for having both Contacts.swift `add_relay` & `state.pool.add_relay`? 

Those two achieve subtly different things:
- The `add_relay` from `RelayPool` (`state.pool`) is concerned about the actual relay WebSocket connection. It adds the relay to the class that handles WebSockets, so that a connection can be established
- The `add_relay` from Contacts.swift is concerned about the user's contact list. On Nostr, users publish the list of relays they normally connect to, to help the network with clues on where to find their notes, and to help other clients to know the preferred relays of a given user (In case they login to a different client)

> I'm also not sure edge case ensuring `relays.index(forKey: relay) == nil` achieves?

It makes sure the relay is not already on the user's contact list. If it already is, there is no need to publish a new relay list to the network.

> I personally don't see a path where this should ever succeed since we are adding the relay to the pool before we are calling Contacts.swift `add_relay` (which expects the relay to not have been added to the pool).

Based on my answers above, the happy path is when the function passes those guard statements and successfully adds the relay to both the connection pool, and sends out a new relay list to the network


I hope it helps, please let me know if you have any other questions!

**Comment by fishcharlie** [comment]
TLDR: I think I fixed this. Submitted a patch: https://groups.google.com/a/damus.io/g/patches/c/_SYF79OdUAY

---

Ok I think I figured out what is going on here. If a `kind:3` message content value is empty then relays will include the relay that was just added:

https://github.com/damus-io/damus/blob/84cfeb16048c6387a38eaac5d1fbf1a1d1762a13/damus/Models/Contacts.swift#L204-L209

Which will cause that to return nil, since it thinks it already exists in `kind:3`. Even tho it doesn't.

https://github.com/damus-io/damus/blob/84cfeb16048c6387a38eaac5d1fbf1a1d1762a13/damus/Models/Contacts.swift#L234-L236

When adding a relay it assumes `ensure_relay_info` will only return relays in the `kind:3` message. However, it can also return the list of all the relays.

It looks like `ensure_relay_info` reverts to returning the list of all the relays to handle the case of deleting a relay. Which makes sense because if the content is empty, and you are deleting a relay, you want the full list of relays (even if `kind:3` is empty)

Whereas for adding a relay, it returns nil if the relay already exists in that list (which it was added previously).

Possible solutions (off the top of my head, there might be more):

1. Create two methods (one for delete and one for add relays) ie. `ensure_relay_info` & `ensure_all_relay_info` where the first uses `?? [:]` and the latter uses the existing `?? make_contact_relays(relays)`. Then we can just use `ensure_all_relay_info` for deleting only and `ensure_relay_info` for adding only.
    - The issue with this is when you add a relay, and `content` is empty, it will only send that 1 new relay for `kind:3`. It won't include existing relays.
2. We could change the guard statement for adding relays to be: `ev.content.isEmpty || relays.index(forKey: relay) == nil`. This would bypass the `return nil` if `ev.content.isEmpty`.
    - The downside with this method is there is still an edge case where if the relay already exists in your `kind:3` list, it will fail to properly add the relay. To prevent this, we will have to move a lot of that code into an `if` block instead of guarding against it. That way the rest of the code can continue without issue.

I have decided to implement solution 2. And submitted a patch here: https://groups.google.com/a/damus.io/g/patches/c/_SYF79OdUAY

**Comment by jb55** [comment]
On Sun, Jan 07, 2024 at 10:18:58AM -0800, Charlie Fish wrote:
>TLDR: I think I fixed this. Submitted a patch: https://groups.google.com/a/damus.io/g/patches/c/_SYF79OdUAY
>
>---
>
>Ok I think I figured out what is going on here. If a `kind:3` message content value is empty then relays will include the relay that was just added:
>
>https://github.com/damus-io/damus/blob/84cfeb16048c6387a38eaac5d1fbf1a1d1762a13/damus/Models/Contacts.swift#L204-L209
>
>Which will cause that to return nil, since it thinks it already exists in `kind:3`. Even tho it doesn't.
>
>https://github.com/damus-io/damus/blob/84cfeb16048c6387a38eaac5d1fbf1a1d1762a13/damus/Models/Contacts.swift#L234-L236
>
>When adding a relay it assumes `ensure_relay_info` will only return
>relays in the `kind:3` message. However, it can also return the list of
>all the relays.
>
>It looks like `ensure_relay_info` reverts to returning the list of all
>the relays to handle the case of deleting a relay. Which makes sense
>because if the content is empty, and you are deleting a relay, you want
>the full list of relays (even if `kind:3` is empty)
>
>Whereas for adding a relay, it returns nil if the relay already exists in that list (which it was added previously).

This is a good explanation. It would be great if this was in the actual
commit message, this context would have been great when I was reviewing
the patch. Just a heads up for next time!

**Comment by fishcharlie** [comment]
> It would be great if this was in the actual
commit message, this context would have been great when I was reviewing
the patch. Just a heads up for next time!

@jb55 I will make that adjustment for next time. Thanks!
